### PR TITLE
feat: Suppress "Unauthorized" errors

### PR DESCRIPTION
**Pull request by Tomasz-Smelcerz-SAP**
**Description**

This PR intention is to prevent from etcd outages which we observe during secret rotation "peaks", when many SKR secrets are rotated in a short time.
Because of secrets rotation, the Lifecycle-Manager encounters a lot of connection errors and is producing many k8s Events objects.
We found out that some of these events are very big. Reducing the size of events may reduce the stress on etcd to an acceptable level.

Details:
When the secret is rotated, the client we use to access SKR returns an "Unauthorized" error. If this happens on a `patch` request, the error message contains a json-serialized copy of the object being patched.
Because we're putting these error messages into related k8s Event, this results in a very big `Event.message` field. In some cases it exceeds 90KiB.
This pull requests detects "Unauthorized" errors produced by go-client and replaces them with our custom error.
In addition, when all resources in a single SSA operation error-out with "Unauthorized", instead of concatenating them into one "big" error, they are all replaced by just a single custom "Unauthorized" error.
This reduces `Event.message` contents from over 90KiB to about 60 bytes - in the extreme case.

Side effects: We lose precise info which objects were patched when `Unauthorized` error occurred. But for now it looks we don't need this - we assume`Unauthorized` means that the runtime secret is not valid anymore and **every** operation on that runtime will return `Unauthorized`.
 
Changes proposed in this pull request:

- Replace client-go `Unauthorized` error with a custom one
- Replace all the errors produced by a single SSA batch operation with a single custom `Unauthorized` error, if all the individual errors are `Unauthorized`.

### CODE REVIEW

**Comment by Tomasz-Smelcerz-SAP** [comment]
Some logs from my local k3d cluster that shows the events "before" and "after" this PR:
Note: The error is custom - it's NOT the client-go error, but it is enough to show the structure of the `Event.message` field
**BEFORE** (all the errors are reported individually)
```
      "kind": "Event",
      "lastTimestamp": "2024-01-08T16:24:04Z",
      "message": "(combined from similar events): patch for configmaps/template-operator-manager-config failed: My Unauthorized\npatch for roles/template-operator-leader-e
lection-role failed: My Unauthorized\npatch for clusterroles/template-operator-metrics-reader failed: My Unauthorized\npatch for clusterroles/template-operator-proxy-role
failed: My Unauthorized\npatch for namespaces/template-operator-system failed: My Unauthorized\npatch for serviceaccounts/template-operator-controller-manager failed: My U
nauthorized\npatch for clusterrolebindings/template-operator-manager-rolebinding failed: My Unauthorized\npatch for services/template-operator-metrics-service failed: My U
nauthorized\npatch for clusterroles/template-operator-manager-role failed: My Unauthorized\npatch for deployments/template-operator-controller-manager failed: My Unauthori
zed\npatch for customresourcedefinitions/samples.operator.kyma-project.io failed: My Unauthorized\npatch for clusterrolebindings/template-operator-proxy-rolebinding failed
: My Unauthorized\npatch for rolebindings/template-operator-leader-election-rolebinding failed: My Unauthorized\nServerSideApply failed (after 45.936597ms)",
```

**AFTER** (all the errors are `Unauthorized`, so they are combined to a single one)
```
      "kind": "Event",
      "lastTimestamp": "2024-01-08T16:54:53Z",
      "message": "Unauthorized\nServerSideApply failed (after 33.213685ms)",
```

File: internal/declarative/v2/ssa.go [hunk 1]
```diff
@@ -61,13 +64,34 @@ func (c *ConcurrentDefaultSSA) Run(ctx context.Context, resources []*resource.In
 	ssaFinish := time.Since(ssaStart)
 
 	if errs != nil {
-		errs = append(errs, fmt.Errorf("%w (after %s)", ErrServerSideApplyFailed, ssaFinish))
+		summaryErr := fmt.Errorf("%w (after %s)", ErrServerSideApplyFailed, ssaFinish)
+		if c.allUnauthorized(errs) {
+			return errors.Join(ErrClientUnauthorized, summaryErr)
+		}
+		errs = append(errs, summaryErr)
```
[the remark this answers is not in the document]
Comment (Tomasz-Smelcerz-SAP) [reply internal/declarative/v2/ssa.go:71]: It already works this way.
The replacement of go-client "Unauthorized" to our custom `ErrClientUnauthorized` in not conditional, it happens **always** in the ssa.go line 129: https://github.com/kyma-project/lifecycle-manager/pull/1222/files#diff-3faec4d40d2495f3366dd0e0a1abd8797165a043caca8522b326dec3ebe63d3dR129

Here I am just detecting if **all** of the errors are the same (`ErrClientUnauthorized`) and if so, I am collapsing these to just a single one. You may argue that we should "collapse" all `ErrClientUnauthorized` errors in the slice to just one, regardless of how they are mixed with other errors. But this means more complex logic and for now I am not convinced if this logic will ever be used. Let's proceed step-by-step here :)

[the remark this answers is not in the document]
Comment (jeremyharisch) [reply internal/declarative/v2/ssa.go:71]: Ahh sorry, when writing this comment, I missed the code you are referencing. You are correct! Thanks for clarification 👍🏻

**Pull request merged**